examples/stable_diffusion_v2/train_controlnet.py

```diff
@@ -0,0 +1,396 @@
+"""
+Train Controlnet
+"""
+import argparse
+import importlib
+import logging
+import os
+import shutil
+
+from ldm.data.dataset import MODE, build_controlnet_dataset
+from ldm.modules.logger import set_logger
+from ldm.modules.lora import inject_trainable_lora
+from ldm.modules.train.callback import EvalSaveCallback, OverflowMonitor
+from ldm.modules.train.checkpoint import resume_train_network
+from ldm.modules.train.ema import EMA
+from ldm.modules.train.lr_schedule import create_scheduler
+from ldm.modules.train.optim import build_optimizer
+from ldm.modules.train.parallel_config import ParallelConfig
+from ldm.modules.train.tools import parse_with_config, set_random_seed
+from ldm.modules.train.trainer import TrainOneStepWrapper
+from ldm.util import count_params, is_old_ms_version, str2bool
+from omegaconf import OmegaConf
+
+import mindspore as ms
+from mindspore import Model, context, load_checkpoint, load_param_into_net
+from mindspore.communication.management import get_group_size, get_rank, init
+from mindspore.nn.wrap.loss_scale import DynamicLossScaleUpdateCell
+from mindspore.train.callback import LossMonitor, TimeMonitor
+
+os.environ["HCCL_CONNECT_TIMEOUT"] = "6000"
+SD_VERSION = os.getenv("SD_VERSION", default="2.0")
+
+logger = logging.getLogger(__name__)
+
+
+def init_env(args):
+    set_random_seed(args.seed)
+
+    # ms.set_context(mode=context.GRAPH_MODE)  # needed for MS2.0
+
+    if args.use_parallel:
+        init()
+        device_id = int(os.getenv("DEVICE_ID"))
+        device_num = get_group_size()
+        ParallelConfig.dp = device_num
+        rank_id = get_rank()
+        args.rank = rank_id
+        logger.debug("Device_id: {}, rank_id: {}, device_num: {}".format(device_id, rank_id, device_num))
+        context.reset_auto_parallel_context()
+        context.set_auto_parallel_context(
+            parallel_mode=context.ParallelMode.DATA_PARALLEL,
+            # parallel_mode=context.ParallelMode.AUTO_PARALLEL,
+            gradients_mean=True,
+            device_num=device_num,
+        )
+    else:
+        device_num = 1
+        device_id = int(os.getenv("DEVICE_ID", 0))
+        rank_id = 0
+        args.rank = rank_id
+
+    context.set_context(
+        # mode=context.PYNATIVE_MODE, #context.GRAPH_MODE
+        mode=context.GRAPH_MODE,
+        device_target="Ascend",
+        device_id=device_id,
+        max_device_memory="30GB",  # TODO: why limit?
+        # pynative_synchronize=True, # to debug
+        # save_graphs=3,
+        # save_graphs_path="output0920/try_silu_output_graph_1022/graphs3",
+    )
+    ms.set_context(ascend_config={"precision_mode": "allow_fp32_to_fp16"})  # Only effective on Ascend 901B
+
+    return rank_id, device_id, device_num
+
+
+def build_model_from_config(config):
```

Comment:
I think it's better to reuse `build_model_from_config`, `get_obj_from_str`, `load_pretrained_model`, and `load_pretrained_model_vae_unet_cnclip` from `train_text_to_image.py` since they are identical.

Reply:
I am afraid this file's functions would change over time.

Reply:
That's the point. If there's a bug / API change in any of those functions, we will need to update all instances of these functions throughout the project, which will complicate code maintenance.